### PR TITLE
exclude kustomization files from KRM check

### LINKED ISSUE #2388: Support Kustomize
### Expected behavior

Running functions on packages with `kustomization.yaml` files works

### Actual behavior

```bash
❯ kpt fn eval --image gcr.io/kpt-fn/apply-setters:v0.1
[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[FAIL] "gcr.io/kpt-fn/apply-setters:v0.1"
Error: input resource list must contain only KRM resources: kustomization.yaml: resource must have `metadata.name` 
```

### Information

```bash
❯ kpt version
1.0.0-beta.1
```
Adding `metadata.name` to the Kustomize file resolves the error, but `metadata` is not in the Kustomize spec 

### Steps to reproduce the behavior

Run `kpt fn eval --image gcr.io/kpt-fn/apply-setters:v0.1` a `kustomize.yaml` file

### PR DESCRIPTION
**Pull request by droot**
This PR excludes kustomization (file or resources) from the KRM check.

fixes #2388

### CODE REVIEW

**Comment by mortent** [comment]
How does the rest of kpt handle KRM files without `metadata.name`? In particular we should probably make sure this also works for packaging. Live seems somewhat less relevant here since resources without `metadata.name` can't be applied anyway, but we should probably verify the error message in these situations.

File: pkg/api/kptfile/v1/validation.go [hunk 1]
```diff
@@ -27,6 +27,13 @@ import (
 	"sigs.k8s.io/kustomize/kyaml/yaml"
 )
 
+const (
+	// constants related to kustomize
+	kustomizationKind     = "Kustomization"
+	kustomizationAPIGroup = "kustomize.config.k8s.io"
+	kustomizationFile     = "kustomization.yaml"
```
Comment (natasha41575) [inline pkg/api/kptfile/v1/validation.go:34]: There are actually 3 valid kustomization file names - https://github.com/kubernetes-sigs/kustomize/blob/59c410a70af15ed330cfd5292b1a642692a7b773/api/konfig/general.go#L10-14

`kustomization.yaml` is the most common, but I wanted to point this out in case you wanted to exclude them all.

**Comment by droot** [comment]
>  In particular we should probably make sure this also works for packaging.

I tested the `pkg update` workflow with [an example pkg](https://github.com/droot/example-pkg-nginx) with kustomization files that doesn't include `metadata.name` (also tested one with no apiVersion and kind field), the pkg update was able to merge fine.

I also tested with an edge case where upstream has metadata.name field and pkg on disk doesn't, `pkg update` results into two resources (duplication), but that is expected behavior with any resource. [So either user will have `metadata.name` or they won't in both upstream/downstream and it works fine for both the cases]

In general, this is help users who are using some upstream package that includes kustomizations that don't have apiVersion,kind and metadata.name. If we ask them to add `metadata.name`, then `pkg update` will run into duplicate resource problem. They also don't want to exclude these kustomizations because they want to run `fn eval` or `fn render` in some form on these.

And they are either running `kustomize build | kubectl apply -` or `kubectl apply -k` to finally apply the resources. 

> Live seems somewhat less relevant here since resources without metadata.name can't be applied anyway, but we should probably verify the error message in these situations.

I think we are tracking that as part of this https://github.com/GoogleContainerTools/kpt/issues/2305

File: pkg/api/kptfile/v1/validation.go (hunk 1 above)
Comment (droot) [reply pkg/api/kptfile/v1/validation.go:34]: Thanks @natasha41575 . I will try to use the constants from api pkg itself.

**Comment by phanimarupaka** [comment]
> > In particular we should probably make sure this also works for packaging.
> 
> I tested the `pkg update` workflow with [an example pkg](https://github.com/droot/example-pkg-nginx) with kustomization files that doesn't include `metadata.name` (also tested one with no apiVersion and kind field), the pkg update was able to merge fine.
> 

But the example pkg has `metadata.name` field.

> I also tested with an edge case where upstream has metadata.name field and pkg on disk doesn't, `pkg update` results into two resources (duplication), but that is expected behavior with any resource. [So either user will have `metadata.name` or they won't in both upstream/downstream and it works fine for both the cases]
> 
> In general, this is help users who are using some upstream package that includes kustomizations that don't have apiVersion,kind and metadata.name. If we ask them to add `metadata.name`, then `pkg update` will run into duplicate resource problem. They also don't want to exclude these kustomizations because they want to run `fn eval` or `fn render` in some form on these.
> 
> And they are either running `kustomize build | kubectl apply -` or `kubectl apply -k` to finally apply the resources.
> 
> > Live seems somewhat less relevant here since resources without metadata.name can't be applied anyway, but we should probably verify the error message in these situations.
> 
> I think we are tracking that as part of this #2305

File: pkg/api/kptfile/v1/validation.go [hunk 2]
```diff
@@ -169,7 +176,22 @@ func GetValidatedFnConfigFromPath(pkgPath types.UniquePath, configPath string) (
 	return nodes[0], nil
 }
 
+func AreKRM(nodes []*yaml.RNode) error {
```
Comment (phanimarupaka) [inline pkg/api/kptfile/v1/validation.go:179]: nit: godoc

**Comment by phanimarupaka** [comment]
> > In particular we should probably make sure this also works for packaging.
> 
> I tested the `pkg update` workflow with [an example pkg](https://github.com/droot/example-pkg-nginx) with kustomization files that doesn't include `metadata.name` (also tested one with no apiVersion and kind field), the pkg update was able to merge fine.
> 

One other case I would test is, `metadata` exists with labels/annotations but `name` and `namespace` doesn't exist. The merge comment will be `/` I suppose. Then try to test the merging. I think we need to have test cases in packaging as well for this as the current cases are based on the assumption that metadata.name exists.

> I also tested with an edge case where upstream has metadata.name field and pkg on disk doesn't, `pkg update` results into two resources (duplication), but that is expected behavior with any resource. [So either user will have `metadata.name` or they won't in both upstream/downstream and it works fine for both the cases]
> 
> In general, this is help users who are using some upstream package that includes kustomizations that don't have apiVersion,kind and metadata.name. If we ask them to add `metadata.name`, then `pkg update` will run into duplicate resource problem. They also don't want to exclude these kustomizations because they want to run `fn eval` or `fn render` in some form on these.
> 
> And they are either running `kustomize build | kubectl apply -` or `kubectl apply -k` to finally apply the resources.
> 
> > Live seems somewhat less relevant here since resources without metadata.name can't be applied anyway, but we should probably verify the error message in these situations.
> 
> I think we are tracking that as part of this #2305

File: pkg/api/kptfile/v1/validation.go (hunk 2 above)
Comment (droot) [reply pkg/api/kptfile/v1/validation.go:180]: Done

**Review by droot — COMMENTED** [review]
Addressed the nit.

**Comment by droot** [comment]
> One other case I would test is, metadata exists with labels/annotations but name and namespace doesn't exist. The merge comment will be / I suppose. Then try to test the merging. I think we need to have test cases in packaging as well for this as the current cases are based on the assumption that metadata.name exists.

Tested this out, I see `metadata: # kpt-merge: /` after the `pkg update` and merge went fine, so looks like we are good. Thanks for the tip.

**Comment by phanimarupaka** [comment]
> > One other case I would test is, metadata exists with labels/annotations but name and namespace doesn't exist. The merge comment will be / I suppose. Then try to test the merging. I think we need to have test cases in packaging as well for this as the current cases are based on the assumption that metadata.name exists.
> 
> Tested this out, I see `metadata: # kpt-merge: /` after the `pkg update` and merge went fine, so looks like we are good. Thanks for the tip.

Great. Thanks for checking. Please add the tests here confirming the behavior. It should be very easy. 

https://github.com/GoogleContainerTools/kpt/blob/main/internal/util/merge/merge3_test.go#L191

**Review by phanimarupaka — APPROVED** [review]

**Comment by droot** [comment]
> Please add the tests here confirming the behavior. It should be very easy.

Added the merge unit test in the latest commit: https://github.com/GoogleContainerTools/kpt/pull/2409/commits/52a38005ed533674691c62010299428164a9ff2c

I see some difference in behavior in manual vs unit test, in the manual tests, `kpt-merge: /` was being added by `pkg update` or even `pkg get` while in the unit test I don't see that being added. Do you know what must be happening here ?

**Comment by phanimarupaka** [comment]
# kpt-merge: <name>/<namespace> is added by `kpt pkg get`. In this you are checking the merge behavior if there is no name and namespace field.

**Comment by droot** [comment]
> `# kpt-merge: <name>/<namespace>` is added by `kpt pkg get`. In this you are checking the merge behavior if there is no name and namespace field.

Thanks for confirming this. I am happy with the tests, so merging this.

**Review by mortent — APPROVED** [review]
